### PR TITLE
Register the queue cleanup callback to vm_id handle callbacks 

### LINKED ISSUE #86: Shared data and shared queue leak on VM change
looks like this repo is the right place to solve

ref: 
- https://github.com/envoyproxy/envoy/issues/13835
- https://github.com/istio/istio/issues/27783

### PR DESCRIPTION
**Pull request by mathetake**
resolves #86, envoyproxy/envoy#13835, istio/istio#27783

### CODE REVIEW

**Comment by mathetake** [comment]
something broken https://github.com/envoyproxy/envoy/pull/14315 😞  will take a look tomorrow

**Comment by PiotrSikora** [comment]
> something broken [envoyproxy/envoy#14315](https://github.com/envoyproxy/envoy/pull/14315) 😞 will take a look tomorrow

I'm afraid that's because of changes in #99 (notably, initialization-order-fiasco because `global_shared_data` gets initialized before `global_vm_id_handle_callbacks` on which it depends due to `registerVmIdHandleCallback`).

**Comment by mathetake** [comment]
ooops, that makes sense. Thanks

**Comment by mathetake** [comment]
so now only thing to solve is Windows build failure...

**Comment by mathetake** [comment]
wait until #115 is merged

File: src/shared_queue.h
Comment (mathetake) [inline src/shared_queue.h:55]: I've started thinking that why not just use `hash(vm_id, queue_name) mod (uint32 max)` as queue tokens. @PiotrSikora thoughts?

Comment (PiotrSikora) [reply src/shared_queue.h:55]: Perhaps? But then you need to worry about collisions, and users see weird queue IDs instead of incrementing numbers.

Comment (mathetake) [reply src/shared_queue.h:55]: > users see weird queue IDs instead of incrementing numbers.

right. I leave as it is for now.

File: src/shared_queue.h
Comment (PiotrSikora) [inline src/shared_queue.h:57]: Is there any reason why this is `std::map` instead of `std::ordered_map`?

**Review by PiotrSikora — APPROVED** [review]

File: src/shared_queue.h
Comment (mathetake) [reply src/shared_queue.h:57]: hah, no reason. just a mistake. give me a sec

File: src/shared_queue.cc
Comment (PiotrSikora) [inline src/shared_queue.cc:75]: Nit: What's the reason for the change from automatic to manual calculation of the hash (using `queueKeyHash` instead of allowing `std::unordered_map` to calculate hash of the key)?

Comment (mathetake) [reply src/shared_queue.cc:75]: I thought it was necessary somehow, but I realize now it's not. Do you want me to roll back? I'm ok with that.

Comment (PiotrSikora) [reply src/shared_queue.cc:75]: I think it makes things a bit more confusing (what's this special hash? where is `size_t` coming from?), so it's not an improvement in my mind, but if you prefer to leave it, then we can leave it. You're the one who's going to maintain this anyway :)

Comment (mathetake) [reply src/shared_queue.cc:75]: > a bit more confusing 

That's true. I don't have preference so I will revert the change :+1:

File: src/shared_queue.cc
Comment (PiotrSikora) [inline src/shared_queue.cc:43]: Nit: `for (auto ...) {`

File: src/shared_queue.cc
Comment (PiotrSikora) [inline src/shared_queue.cc:88]: Instead of searching `vm_queue_keys_` first, you could move that code into the `if (it.second) {` block, since we know that the `vm_queue_keys_` already exists if it evaluates to `false`.

File: src/shared_queue.cc
Comment (PiotrSikora) [inline src/shared_queue.cc:101]: This seems like an unnecessary change. Compiler is going to optimize it for you anyway (at least, I hope so), and it makes the code less readable.

File: test/shared_queue_test.cc
Comment (PiotrSikora) [inline test/shared_queue_test.cc:40]: `100` seems a bit too much here.

File: test/shared_queue_test.cc
Comment (PiotrSikora) [inline test/shared_queue_test.cc:131]: Could you split this into 2 separate `for` loops? Having `i++` in the middle of the loop doesn't help with readability of this code.

File: test/shared_queue_test.cc
Comment (PiotrSikora) [inline test/shared_queue_test.cc:138]: Same here.

File: src/shared_queue.cc
Comment (mathetake) [reply src/shared_queue.cc:88]: right. Thanks

File: src/shared_queue.cc
Comment (PiotrSikora) [inline src/shared_queue.cc:75]: Nit: leave the old code, i.e. `auto key = std::make_pair(std::string(vm_id), std::string(queue_name))`, and assign `vid` inside `if` statement only if needed.

**Review by PiotrSikora — APPROVED** [review]

**Comment by PiotrSikora** [comment]
Thanks!

**Pull request merged**
